docs/proposals/20221003-In-place-propagation-of-Kubernetes-objects-only-changes.md

```diff
@@ -0,0 +1,220 @@
+
+Note: as of today topology controllers already propagates ClusterClass and Cluster Topology metadata changes in place when possible
+in order to avoid unnecessary template rotation with the consequent Machine rollout; we do not foresee changes to this logic
+in order to support above propagation rules.
+
+#### Simplified Metadata propagation from MachineDeployment to MachineSet
+
+We are going to drop current behaviour inherithed from Kubernetes's Deployment controller, that computes top level labels in MachineSet
+using a different source from the one used for annotations. Instead we propose a more straight forward propagation:
+
+1. Top level metadata from MachineDeployment propagates to top level metadata in MachineSet.
+2. spec.template.metadata from MachineDeployment propagates to spec.template.metadata in MachineSet.
+
+In order to prevent unnecessary rollouts, the above propagation should happen in place (see [in place propagation](#in-place-propagation));
+same considerations apply to metadata propagation from MachineSet to machines.
+
+#### In Place Metadata propagation from KubeadmControlPlane to Machines
+
+In order to prevent unnecessary rollouts, metadata propagation from KubeadmControlPlane to Machines should happen in place (see [in place propagation](#in-place-propagation)).
+
+#### Label Sync Between Machine and underlying Kubernetes Nodes
+
+As discussed in [Label Sync Between Machine and underlying Kubernetes Nodes](link to the proposal as soon as it merge) we are propagating from
+the Machine to corresponding Kubernetes Node only labels with a well known prefix or a well known domain.
+
+### Propagation of fields impacting only Kubernetes objects or controller behaviour
+
+Apart from labels and annotations, there are also other fields that flow downs from ClusterClass to KubeadmControlPlane/MachineDeployments and
+ultimately to Machines.
+
+Some of them can be considered like labels and annotations, because they have impacts only only Kubernetes objects or controller behaviour, but
+not on the Machine itself. Examples are `MinReadySeconds`, `NodeDrainTimeout`, `NodeVolumeDetachTimeout`, `NodeDeletionTimeout`.
+
+Propagation of changes to those fields will be implementated using the same [in place propagation](#in-place-propagation) mechanism implemented
+for metatada .
+
+### In place propagation
+
+With in place propagation we are referring to a mechanism that updates existing Kuberentes objects, like MachineSets or Machines, instead of
+creating a new instance with the updated info and deleting the current Kubernetes object.
+
+The main benefits of this approach is that it prevents unnecessary rollouts of the corresponding infrastructure, with the consequent creation/
+deletion of a Kubernetes node and drain/scheduling ofworkloads hosted on the machine being deleted.
+
+**Important!** in place propagation of changes as defined above applies only to metadata changes or to fields impacting only Kubernetes objects
```

Comment:
may be we should agree on a name for this definition and add it to the glossary so we refer to the fields that qualify consistently. E.g

Elastic fields: Fields which changes would only impact Kubernetes objects or/and controller behaviour but they won't mutate in any way provider infrastructure nor the software running on it. Elastic fields are propagated in place by CAPI controllers to avoid the more elaborated mechanics of a replace rollout. These includes metadata, MinReadySeconds, NodeDrainTimeout, NodeVolumeDetachTimeout and NodeDeletionTimeout but are not limited to be expanded in the future.

Reply:
I'm +1 to add a definition to the glossary, let's see if we can find consensus on the name/definition

Reply:
Definitely a good idea to define this and then just use it going forward. 

I'm fine with the definition, not sure about the name.

But I don't have a good idea, maybe "In-place mutable fields" would be very very explicit

Reply:
added definition to the glossary in the proposal and to the glossary in the book